### PR TITLE
Refactored logic in sstore_cost

### PR DESCRIPTION
**Pull request by mw2000**
Refactored the sstore_cost function to be more readable and maintainable as mentioned in #830. Broke it down into smaller functions and removed the nesting by using return statements.

### CODE REVIEW

File: crates/interpreter/src/gas/calc.rs [hunk 1]
```diff
@@ -193,50 +193,56 @@ pub fn sstore_cost<SPEC: Spec>(
...
-    // In EIP-2929 we charge extra if the slot has not been used yet in this transaction
+
     if SPEC::enabled(BERLIN) && is_cold {
         Some(gas_cost + COLD_SLOAD_COST)
     } else {
         Some(gas_cost)
     }
 }
 
+fn calculate_istanbul_gas_cost(
```
Comment (rakita) [inline crates/interpreter/src/gas/calc.rs:220]: renaming it to `istanbul_sstore_cost` feels better. and put `#[inline(always]` there

File: crates/interpreter/src/gas/calc.rs [hunk 2]
```diff
@@ -193,50 +193,56 @@ pub fn sstore_cost<SPEC: Spec>(
     gas: u64,
     is_cold: bool,
 ) -> Option<u64> {
```
Comment (rakita) [inline crates/interpreter/src/gas/calc.rs:195]: This whole function could be seen as.
```
if SPEC::enabled(BERLIN) {

} else if  SPEC::enabled(ISTANBUL) {

} else {

}
```

Where berlin path would call instanbul function as they have similar functionality.

File: crates/interpreter/src/gas/calc.rs (hunk 1 above)
Comment (mw2000) [reply crates/interpreter/src/gas/calc.rs]: Resolved this

File: crates/interpreter/src/gas/calc.rs (hunk 2 above)
Comment (mw2000) [reply crates/interpreter/src/gas/calc.rs:195]: Resolved, I've left the 1706 check for Istanbul outside the main if-else condition statement though

File: crates/interpreter/src/gas/calc.rs [hunk 3]
```diff
@@ -193,47 +193,57 @@ pub fn sstore_cost<SPEC: Spec>(
...
+}
+
+#[inline(always)]
+fn istanbul_sstore_cost(
+    original: U256,
+    current: U256,
+    new: U256,
+    gas_sload: u64,
+    gas_sstore_reset: u64,
+) -> u64 {
```
Comment (rakita) [inline crates/interpreter/src/gas/calc.rs:230]: Very readable!

**Review by rakita — CHANGES_REQUESTED** [review]
left a few more comments, but those should be it

File: crates/interpreter/src/gas/calc.rs [hunk 4]
```diff
@@ -193,47 +193,57 @@ pub fn sstore_cost<SPEC: Spec>(
...
+    } else if original == current && original == U256::ZERO {
+        SSTORE_SET
+    } else if original == current {
+        gas_sstore_reset
+    } else {
+        gas_sload
+    }
+}
+
+fn non_istanbul_sstore_cost(current: U256, new: U256, gas_sstore_reset: u64) -> u64 {
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/interpreter/src/gas/calc.rs:242]: done

File: crates/interpreter/src/gas/calc.rs [hunk 5]
```diff
@@ -193,47 +193,57 @@ pub fn sstore_cost<SPEC: Spec>(
...
-                }
-            } else {
-                gas_sload
-            }
+            Some(gas_cost)
         }
+    } else if SPEC::enabled(ISTANBUL) {
+        // Istanbul logic
+        let (gas_sload, gas_sstore_reset) = (sload_cost::<SPEC>(is_cold), SSTORE_RESET);
+        Some(istanbul_sstore_cost(original, current, new, gas_sload, gas_sstore_reset))
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/interpreter/src/gas/calc.rs:215]: done

**Comment by rakita** [comment]
Close in https://github.com/bluealloy/revm/pull/974

**Pull request closed without merging**